x/archivedb/batch.go

```diff
@@ -0,0 +1,124 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package archivedb
+
+import "github.com/ava-labs/avalanchego/database"
+
+// A thin wrapper on top of the database.Batch that also keeps track of the height.
+//
+// It takes any key and wraps it with dbKey. It is possible to construct many
+// batches but at Write() time, the height of the batch to be written to the
+// database must be the following height from the last written to the database,
+// otherwise an error will be thrown
+type dbBatchWithHeight struct {
+	db     *archiveDB
+	ops    map[string]database.BatchOp
+	size   int
+	height uint64
+	batch  database.Batch
+}
+
+func newBatchWithHeight(db *archiveDB, height uint64) *dbBatchWithHeight {
+	batch := db.rawDB.NewBatch()
+	ops := make(map[string]database.BatchOp)
+	size := 0
+	return &dbBatchWithHeight{
+		db,
+		ops,
+		size,
+		height,
+		batch,
+	}
+}
+
+// Returns the height for this Batch
+func (c *dbBatchWithHeight) Height() uint64 {
+	return c.height
+}
+
+// Writes the changes to the database
+func (c *dbBatchWithHeight) Write() error {
+	c.db.lock.Lock()
+	defer c.db.lock.Unlock()
+
+	if c.db.currentHeight+1 != c.height {
+		return ErrInvalidBatchHeight
+	}
+
+	if err := c.Replay(c.batch); err != nil {
+		return err
+	}
+	if err := c.batch.Write(); err != nil {
+		return err
+	}
+	c.db.currentHeight = c.height
+
+	return nil
+}
+
+// Delete any previous state that may be stored in the database
+func (c *dbBatchWithHeight) Delete(key []byte) error {
+	rawKey := newDBKey(key, c.height)
```

Comment:
I feel like the `key` argument would actually make more sense named `rawKey` than the way it is now.